### PR TITLE
avm2: Remove some dynamic-dispatch TObject logic

### PR DESCRIPTION
**Pull request by Lord-McSweeney**
All Object variants must now implement a method `gc_base`, which returns a `Gc` around `RefLock<ScriptObjectData>` (which will later be changed to just `ScriptObjectData`). Additionally, `TObject` methods that aren't overriden by the object variants are marked as `#[no_dynamic]`, which prevents them getting duplicated by `enum_trait_object`.

### CODE REVIEW

File: core/macros/src/lib.rs [hunk 1]
```diff
@@ -148,3 +166,14 @@ pub fn enum_trait_object(args: TokenStream, item: TokenStream) -> TokenStream {
 
     out.into()
 }
+
+#[proc_macro_attribute]
+pub fn no_dynamic(_args: TokenStream, item: TokenStream) -> TokenStream {
```
Comment (adrian17) [inline core/macros/src/lib.rs:171]: as moulins proposed, this is probably not needed if you just strip the attribute in proc-macro

File: core/src/avm2/object.rs [hunk 2]
```diff
@@ -153,6 +153,17 @@ pub use crate::avm2::object::xml_list_object::{
 pub use crate::avm2::object::xml_object::{xml_allocator, XmlObject, XmlObjectWeak};
 use crate::font::Font;
 
+fn borrow_data<'gc>(gc: Gc<'gc, RefLock<ScriptObjectData<'gc>>>) -> Ref<ScriptObjectData<'gc>> {
```
Comment (adrian17) [inline core/src/avm2/object.rs:156]: why are these needed?

File: core/src/avm2/object/bitmapdata_object.rs [hunk 3]
```diff
@@ -62,6 +61,11 @@ pub struct BitmapDataObjectData<'gc> {
     bitmap_data: Lock<Option<BitmapDataWrapper<'gc>>>,
 }
 
+const _: () = assert!(std::mem::offset_of!(BitmapDataObjectData, base) == 0);
```
Comment (adrian17) [inline core/src/avm2/object/bitmapdata_object.rs:64]: can these be moved to inside the method?
(Maybe even generated by a macro, to hide the repeated `unsafe`? ;D )

File: core/src/avm2/object.rs (hunk 2 above)
Comment (Lord-McSweeney) [reply core/src/avm2/object.rs:156]: I'm getting bad codegen when I remove these, though I'm not sure why, they're getting inlined. These are going to disappear as soon as ScriptObjectData gets interior mutability, though, so I don't think it's too big of a problem.

File: core/src/avm2/object/bitmapdata_object.rs (hunk 3 above)
Comment (Lord-McSweeney) [reply core/src/avm2/object/bitmapdata_object.rs:64]: I think it makes more sense to have them here, as they're referencing the struct immediately above, and the TObject impl isn't right after the `struct AbcObjectData` definition for several object variants.

Comment (adrian17) [reply core/src/avm2/object/bitmapdata_object.rs:64]: My motivation for this was that them being directly above an `unsafe` block explains its safety better, but I'm not hell bent on this :)

File: core/macros/src/lib.rs [hunk 4]
```diff
@@ -62,10 +62,44 @@ pub fn enum_trait_object(args: TokenStream, item: TokenStream) -> TokenStream {
...
                 let method_name = &method.sig.ident;
+
+                let mut is_no_dynamic = false;
+
+                let new_attrs = method
+                    .attrs
+                    .iter()
+                    .filter(|attr| match &attr.meta {
+                        Meta::Path(path) => {
+                            if let Some(ident) = path.get_ident() {
```
Comment (adrian17) [inline core/macros/src/lib.rs:77]: `if path.is_ident("no_dynamic")`?

File: core/macros/src/lib.rs [hunk 5]
```diff
@@ -62,10 +62,44 @@ pub fn enum_trait_object(args: TokenStream, item: TokenStream) -> TokenStream {
...
+                                    true
+                                }
+                            } else {
+                                true
+                            }
+                        }
+                        _ => true,
+                    })
+                    .map(|o| o.clone())
+                    .collect::<Vec<_>>();
```
Comment (adrian17) [inline core/macros/src/lib.rs:94]: instead of copying and reassigning, can we call `method.attrs.retain(|attr| ...)`?

**Review by adrian17 — APPROVED** [review]
Approved, assuming notes are resolved.
I just wish there was a nicer way to wrap/hide/dedupe these `unsafe`s :/

I still think about smth like
`impl_base_cast!(ClassObjectData, ScriptObjectData, base)` that adds both the method and asserts in one go. (but we can leave that for the future)
